MCXW23 zephyr base enablement

AllenZhang-NXP opened the pull request:
soc: mcxw235,mcxw236: add SOC support for MCXW235 and MCXW236
dts: mcxw235,mcxw236: add dts for MCXW235 and MCXW236
boards: frdm_mcxw23: add frdm_mcxw23 board
boards: mcxw23_evk: add mcxw23_evk board

AllenZhang-NXP committed:
soc: mcxw235,mcxw236: add SOC support for MCXW235 and MCXW236

add soc MCX235 and MCXW236 for board frdm_mcxw23

Signed-off-by: Allen Zhang <chunfeng.zhang@nxp.com>

AllenZhang-NXP committed:
dts: mcxw235,mcxw236: add dts for MCXW235 and MCXW236

add dts for device MCXW235 and MCXW236

Signed-off-by: Allen Zhang <chunfeng.zhang@nxp.com>

AllenZhang-NXP committed:
boards: frdm_mcxw23: add frdm_mcxw23 board

enable board support for frdm_mcxw23

Signed-off-by: Allen Zhang <chunfeng.zhang@nxp.com>

AllenZhang-NXP committed:
boards: mcxw23_evk: add mcxw23_evk board

enable board support for mcxw23_evk

Signed-off-by: Allen Zhang <chunfeng.zhang@nxp.com>

nordicjm commented on soc/nxp/mcx/mcxw2x/CMakeLists.txt:
```diff
@@ -0,0 +1,23 @@
+# CMSIS SystemInit allows us to skip enabling clock to SRAM banks via
+# this compiler definition
+if(NOT DEFINED CONFIG_MCXW23X_SRAM_CLOCKS)
+zephyr_compile_definitions(DONT_ENABLE_DISABLED_RAMBANKS=1)
```
2 space indent

AllenZhang-NXP replied:
Fixed

nordicjm commented on soc/nxp/mcx/mcxw2x/Kconfig.defconfig:
```diff
@@ -0,0 +1,31 @@
+	default 60
+
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 32000000
```
get from dts value using dt functions

mmahadevan108 replied:
You can reference this line for an example. 
https://github.com/zephyrproject-rtos/zephyr/blob/main/soc/nxp/imxrt/Kconfig.defconfig#L27
You would need to add `clock-frequency = 32000000` for the sysclk node in the soc dts file

AllenZhang-NXP replied:
Fixed

nordicjm commented on dts/arm/nxp/nxp_mcxw23x_common.dtsi:
```diff
@@ -0,0 +1,288 @@
+		compatible = "nxp,lpc-syscon";
+		reg = <0x0 0x4000>;
+		#clock-cells = <1>;
+		reset: reset {
```
newline line 79, fix in whole PR

nordicjm commented on boards/nxp/frdm_mcxw23/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+	default y
+	depends on FXLS8974_TRIGGER
+
+if BOOTLOADER_MCUBOOT
+choice MCUBOOT_BOOTLOADER_MODE
+	# Board only supports MCUBoot via "upgrade only" method:
+	default MCUBOOT_BOOTLOADER_MODE_OVERWRITE_ONLY
+endchoice
+endif #BOOTLOADER_MCUBOOT
```
```suggestion
if BOOTLOADER_MCUBOOT

choice MCUBOOT_BOOTLOADER_MODE
	# Board only supports MCUBoot via "upgrade only" method:
	default MCUBOOT_BOOTLOADER_MODE_OVERWRITE_ONLY
endchoice

endif # BOOTLOADER_MCUBOOT
```

nordicjm commented on boards/nxp/frdm_mcxw23/doc/frdm_mcxw23.webp:
needs vastly reducing in size/resolution

nordicjm commented on boards/nxp/frdm_mcxw23/frdm_mcxw23_common.dtsi:
```diff
@@ -0,0 +1,201 @@
+		#gpio-cells = <2>;
+		gpio-map-mask = <0xffffffff 0xffffffc0>;
+		gpio-map-pass-thru = <0 0x3f>;
+		gpio-map =	<0 0 &gpio0 16 0>,	/* A0 */
```
space after =, not tab

nordicjm commented on boards/nxp/mcxw23_evk/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+	depends on FXLS8974_TRIGGER
+
+if BOOTLOADER_MCUBOOT
+choice MCUBOOT_BOOTLOADER_MODE
```
apply changes from other commit

AllenZhang-NXP replied:
Fixed

mmahadevan108 commented on boards/nxp/frdm_mcxw23/frdm_mcxw23.yaml:
```diff
@@ -0,0 +1,21 @@
+  - gpio
+  - uart
+  - counter
+  - flash
```
`i2c`?

AllenZhang-NXP committed:
tests: drivers: gpio: add overlay to run test on NXP MCXW23 boards

Support arduino_gpio example for NXP frdm_mcxw23 and mcxw23_evk boards.
Test using tests/drivers/gpio/gpio_basic_api.

Signed-off-by: Allen Zhang <chunfeng.zhang@nxp.com>

AllenZhang-NXP committed:
samples: sensor: enable sensor trigger for frdm_mcxw23

enable sensor trigger.

Signed-off-by: Allen Zhang <chunfeng.zhang@nxp.com>

AllenZhang-NXP commented on boards/nxp/frdm_mcxw23/frdm_mcxw23.yaml:
```diff
@@ -0,0 +1,21 @@
+  - gnuarmemb
+supported:
+  - gpio
+  - uart
```
On going: enable tests/drivers/uart/uart_basic_api with lpuart loopback

mmahadevan108 commented on modules/hal_nxp/mcux/Kconfig.mcux:
```diff
@@ -8,7 +8,8 @@ config HAS_MCUX
 		SOC_FAMILY_NXP_S32 || SOC_FAMILY_NXP_IMXRT || SOC_FAMILY_NXP_RW || \
-		SOC_FAMILY_MCXN || SOC_FAMILY_MCXA || SOC_FAMILY_MCXW || SOC_FAMILY_MCXC
+		SOC_FAMILY_MCXN || SOC_FAMILY_MCXA || SOC_FAMILY_MCXW || SOC_FAMILY_MCXC || \
+		SOC_FAMILY_MCXW2X
```
Why is this need? Would'nt `SOC_FAMILY_MCXW` cover for this SoC?

decsny commented on boards/nxp/mcxw23_evk/mcxw23_evk_common.dtsi:
if it has any similarity with the other mcxw evk, then it should be in the samme folder,  see for example https://github.com/zephyrproject-rtos/zephyr/pull/95577

AllenZhang-NXP replied:
mcxw23_evk and mcxw72_evk boards are totally different.
mcxw23_evk:
Schematic: https://jira.sw.nxp.com/secure/attachment/2327928/mcxw23_evk_v1_1.pdf
Board picture: https://mcuxpresso.nxp.com/mcuxsdk/latest/html/boards/MCX/mcxw23evk/index.html

mcxw72_evk:
Schematic: https://nxp1.sharepoint.com/:b:/r/sites/MCUlibrary/kw47/03_Applications/HARDWARE/Board%20Designs/EVK/Rev%20D/Released%20Files/SPF-91100_D.pdf?csf=1&web=1&e=Kv9MQp
Board picture: https://mcuxpresso.nxp.com/mcuxsdk/latest/html/boards/MCX/mcxw72evk/index.html

decsny commented on boards/nxp/frdm_mcxw23/frdm_mcxw23.yaml:
if this frdm_mcxw board is similar to the other two, we should consider refactoring to combine them all and stop duplicating.

AllenZhang-NXP committed:
soc: mcxw: Move mcxw2x into mcxw and creat mcxw7x subfolder

move mcxw2x into mcxw and created mcxw7x subfolder for mcxw71/mcxw72

Signed-off-by: Allen Zhang <chunfeng.zhang@nxp.com>

AllenZhang-NXP commented on dts/arm/nxp/nxp_mcxw23x_common.dtsi:
The MCXW23 is a rebrand project, most of the peripherals are reused from LPC55S16. So, can not share the dtsi from mcxw7x.

dleach02 commented:
@nordicjm has @AllenZhang-NXP addressed your comments?

AllenZhang-NXP commented on boards/nxp/frdm_mcxw23/frdm_mcxw23_common.dtsi:
```diff
@@ -0,0 +1,204 @@
+
+	leds {
+		compatible = "gpio-leds";
+		red_led: led_0 {
```
Hi @nordicjm, I just fixed the newlines missing issues in frdm_mcxw23_common.dtsi & mcxw23_evk_common.dtsi.

nordicjm commented on soc/nxp/mcx/mcxw/CMakeLists.txt:
```diff
@@ -2,18 +2,34 @@
 
-if (CONFIG_BT OR CONFIG_IEEE802154)
-  zephyr_linker_sources(RAM_SECTIONS sections.ld)
+  if (CONFIG_BT OR CONFIG_IEEE802154)
```
```suggestion
  if(CONFIG_BT OR CONFIG_IEEE802154)
```
no spaces between if and bracket, fix below also

nordicjm reviewed (commented):
minor nit. then looks good

ZhaoxiangJin commented on dts/arm/nxp/nxp_mcxw23x_common.dtsi:
```diff
@@ -0,0 +1,285 @@
+		prescale = <0>;
+	};
+
+	ctimer4: ctimer@2A000 {
```
lower case

ZhaoxiangJin commented on boards/nxp/frdm_mcxw23/frdm_mcxw23_common.dtsi:
```diff
@@ -0,0 +1,214 @@
+		led0 = &red_led;
+		led1 = &green_led;
+		led2 = &blue_led;
+		rtc = &rtc;
```
just curious, why aliases rtc on board level？

ZhaoxiangJin commented on boards/nxp/frdm_mcxw23/frdm_mcxw23_common.dtsi:
```diff
@@ -0,0 +1,214 @@
+			reg = <0x0082000 DT_SIZE_K(440)>;
+		};
+
+		storage_partition: partition@F0000 {
```
lower case

AllenZhang-NXP commented on soc/nxp/mcx/mcxw/Kconfig.soc:
```diff
@@ -23,16 +23,32 @@ config SOC_MCXW727C_CPU1
 	default "MCXW716CMFTA" if SOC_PART_NUMBER_MCXW716CMFTA
 	default "MCXW727CMFTA" if SOC_PART_NUMBER_MCXW727CMFTA
+
+config SOC_PART_NUMBER
```
Fixed. Removed the duplicate 'config SOC_PART_NUMBER'

AllenZhang-NXP commented on tests/drivers/gpio/gpio_basic_api/boards/mcxw23_evk.overlay:
```diff
@@ -0,0 +1,18 @@
+/*
+ * Copyright (c) 2025 NXP
```
Fixed.

AllenZhang-NXP commented on soc/nxp/mcx/mcxw2x/Kconfig.defconfig:
```diff
@@ -0,0 +1,33 @@
+# Copyright 2024 NXP
```
This file is no longer existed after the below commit in this PR is submited:
_soc: mcxw: Move mcxw2x into mcxw and creat mcxw7x subfolder_

hakehuang reviewed (dismissed):
board testing pass

mmahadevan108 commented on soc/nxp/mcx/mcxw/CMakeLists.txt:
I would add the respective subdirectory using `add_subdirectory_ifdef` and add `CMakeLists.txt` in each sub-folder

mmahadevan108 commented on soc/nxp/mcx/mcxw/Kconfig.defconfig:
```diff
@@ -8,13 +8,17 @@ config CORTEX_M_SYSTICK
-	default 96000000 if CORTEX_M_SYSTICK
+	default 96000000 if (CORTEX_M_SYSTICK && (SOC_MCXW727C || SOC_MCXW716C))
+	default $(dt_node_int_prop_int,$(DT_SYSCLK_PATH),clock-frequency) if (CORTEX_M_SYSTICK && (SOC_MCXW236 || SOC_MCXW235))
 
```
Move this to the subfolder.

mmahadevan108 commented on soc/nxp/mcx/mcxw/Kconfig.defconfig:
```diff
@@ -8,13 +8,17 @@ config CORTEX_M_SYSTICK
+	default $(dt_node_int_prop_int,$(DT_SYSCLK_PATH),clock-frequency) if (CORTEX_M_SYSTICK && (SOC_MCXW236 || SOC_MCXW235))
 
 config MCUX_FLASH_K4_API
-	default y
```
Move this to the subfolder

mmahadevan108 commented on soc/nxp/mcx/mcxw/Kconfig.defconfig:
```diff
@@ -64,5 +68,18 @@ config NXP_RF_IMU
+config FLASH_FILL_BUFFER_SIZE
+	default 512
+
+endif # SOC_MCXW236
```
Move

mmahadevan108 commented on soc/nxp/mcx/mcxw/Kconfig.soc:
```diff
@@ -23,16 +23,30 @@ config SOC_MCXW727C_CPU1
 	default "mcxw716c" if SOC_MCXW716C
 	default "mcxw727c" if SOC_MCXW727C
+	default "mcxw236" if SOC_MCXW236
+	default "mcxw235" if SOC_MCXW235
```
Look at `imxrt` for reference. I would retain 
```
config SOC_FAMILY_MCXW
	bool

config SOC_FAMILY
	default "mcxw" if SOC_FAMILY_MCXW
```
Everything else should go into the respective sub-folder's `Kconfig.soc` file

mmahadevan108 commented on soc/nxp/mcx/mcxw/Kconfig:
```diff
@@ -26,3 +26,10 @@ config MCUX_CORE_SUFFIX
+config MCXW23X_SRAM_CLOCKS
+	bool "Clock MCXW23 SRAM banks"
+
+endif # SOC_MCXW236
```
Move the platform specific configs to the subfolder `Kconfig` file

AllenZhang-NXP commented on soc/nxp/mcx/soc.yml:
```diff
@@ -29,6 +29,7 @@ family:
     cpuclusters:
     - name: cpu0
     - name: cpu1
+  - name: mcxw236
```
Added the SOC_SERIES_MCXW2XX and SOC_SERIES_MCXW7XX. Moved the soc's under series.

AllenZhang-NXP commented on soc/nxp/mcx/mcxw/CMakeLists.txt:
```diff
@@ -2,18 +2,7 @@
+add_subdirectory_ifdef(CONFIG_SOC_MCXW716C    mcxw7x)
+add_subdirectory_ifdef(CONFIG_SOC_MCXW727C    mcxw7x)
+add_subdirectory_ifdef(CONFIG_SOC_MCXW236     mcxw2x)
+add_subdirectory_ifdef(CONFIG_SOC_MCXW235     mcxw2x)
```
Updated the CMakeLists.txt.
Added the SOC_SERIES_MCXW2XX and SOC_SERIES_MCXW7XX. 
Rename the subfolders to mcxw2xx annd mcxw7xx.

AllenZhang-NXP commented on soc/nxp/mcx/mcxw/mcxw2x/soc.h:
```diff
@@ -0,0 +1,47 @@
+/* Wrapper Function to deal with SDK differences in power API */
+static inline void EnableDeepSleepIRQ(IRQn_Type irq)
+{
+}
```
Updated soc.h to match https://github.com/zephyrproject-rtos/zephyr/pull/95813.
However, by checking the MCXW23 power driver in _https://github.com/zephyrproject-rtos/hal_nxp/blob/master/mcux/mcux-sdk-ng/devices/MCX/MCXW/MCXW236/drivers/fsl_power.c#L1319_, currently, there's no such API to config the wakeup source before enter into lowpower mode.
We have the API like:
_status_t POWER_EnterDeepSleep(uint32_t excludeFromPd, uint64_t wakeupInterrupts)_
The wakeup interrupts is one of the parameter of API POWER_EnterDeepSleep function.

So, for the moment, I leave NXP_ENABLE_WAKEUP_SIGNAL empty. Once we support the API in power driver,  we can update the soc.h accordingly. I will discuss this further with Jonas who is the power driver of MCXW23.

mmahadevan108 commented on soc/nxp/mcx/mcxw/mcxw2xx/soc.h:
```diff
@@ -0,0 +1,50 @@
+#undef NXP_ENABLE_WAKEUP_SIGNAL
+#undef NXP_DISABLE_WAKEUP_SIGNAL
+#define NXP_ENABLE_WAKEUP_SIGNAL(irqn)
+#define NXP_DISABLE_WAKEUP_SIGNAL(irqn)
```
You don't need to do this, simply `include soc_common.h`, it has the default implementation.

AllenZhang-NXP replied:
OK, got it, I have removed the code here and only include the soc_common.h

hakehuang reviewed (dismissed):
board testing pass

mmahadevan108 committed:
soc: nxp: mcxw7xx: Fix MISRA compliance

Add curly braces around the while block

Signed-off-by: Mahesh Mahadevan <mahesh.mahadevan@nxp.com>

cfriedt commented:
@mmahadevan108 - I'll get it in later this eve.
